Collapse small donut slices into a single "Other" category

dpsutton opened the pull request:
Use viz settings for donut slices

##### Before
<img width="664" alt="image" src="https://user-images.githubusercontent.com/6377293/131147843-28e1e794-453d-401c-8701-55289e6d3e30.png">

##### After
<img width="1042" alt="image" src="https://user-images.githubusercontent.com/6377293/131147875-33117170-8b94-4687-a4c1-afe5683cfdba.png">

#### Reference UI
<img width="1057" alt="image" src="https://user-images.githubusercontent.com/6377293/131147946-8ef7761f-2e6b-468f-98b3-ea72ca3b4fb2.png">

dpsutton committed:
Collapse small donut slices into a single "Other" category

camsaul commented on src/metabase/pulse/render/body.clj:
```diff
@@ -243,19 +243,33 @@
+                        (for [[label value] rows]
+                          [label (if (zero? total)
+                                   (tru "N/A")
+                                   (let [f (DecimalFormat. "###,###.##%")]
```
You should look at the `custom-formatting` Setting's value of `percent` for the format string to use here... see the code @noahmoss wrote for the XLSX stuff for example.

We're not doing it anywhere else but falling back to 
```clj
(java.text.NumberFormat/getPercentInstance (metabase.util.i18n/user-locale))
```

might be better than assuming English-style formatting although that's really a separate question.

camsaul replied:
Actually I think this could show up in viz settings as well (?). So it should be viz settings > `custom-formatting` Setting > default to get the format string

noahmoss replied:
We probably shouldn't do any number formatting in pulses based on user locale, right? Partly because we don't do it on the frontend, but moreso because a pulse could be sent to multiple users with different locales.

dpsutton replied:
yeah seems like the server sets it regardless of where the email goes. The format string gets localized though. `###,###.##%` will use the locale's separators and groupings.

> The characters listed here are used in non-localized patterns. Localized patterns use the corresponding characters taken from this formatter's DecimalFormatSymbols object instead, and these characters lose their special status. Two exceptions are the currency sign and quote, which are not localized. 
https://docs.oracle.com/javase/7/docs/api/java/text/DecimalFormat.html

camsaul commented on src/metabase/pulse/render/body.clj:
```diff
@@ -243,19 +243,33 @@
+  "Process rows with a minimum slice threshold. Collapses any segments below the threshold given as a percentage (the
+  value 25 for 25%) into a single category as \"Other\". "
+  [threshold-percentage rows]
+  (let [total                    (reduce + 0 (map second rows))
```
Are we 100% sure there can't be any `nil` values for the second column? I guess that probably wouldn't make sense if there were

dpsutton replied:
I can make it safer but that's the point of `(common/non-nil-rows x-axis-rowfn y-axis-rowfn rows)`. But you're right it could be safer.

camsaul replied:
sorry I didn't see `non-nil-rows`. Carry on

dpsutton replied:
its further up the call tree. so it could easily/accidentally be removed. it makes sense to throw an `(fnil + 0)` in there

camsaul reviewed (approved):
Overall this looks good to me but we shouldn't hardcode `(DecimalFormat. "###,###.##%")` for formatting percentages... we should use the viz Settings formatting style or the global `custom-formatting` Setting. See what @noahmoss did in the XLSX stuff for example

dpsutton committed:
Locale aware decimal formatting

A followup PR will add all of the formatting options for scalars and
smart scalars so this function is largely temporary and will be
subsumed into a namespace that knows about all formatting options.

dpsutton committed:
empty with no cache [ci nocache]
